runtime/runtime.go

```diff
@@ -1440,6 +1492,356 @@ func (r *interpreterRuntime) newUnsafeRandomFunction(runtimeInterface Interface)
+			var contractType *sema.CompositeType
+			var contractInterfaceType *sema.InterfaceType
+			var declaredName string
+
+			switch {
+			case len(contractTypes) == 1 && len(contractInterfaceTypes) == 0:
+				contractType = contractTypes[0]
+				declaredName = contractType.Identifier
+				deployedType = contractType
+			case len(contractInterfaceTypes) == 1 && len(contractTypes) == 0:
+				contractInterfaceType = contractInterfaceTypes[0]
+				declaredName = contractInterfaceType.Identifier
+				deployedType = contractInterfaceType
+			}
+
+			if deployedType == nil {
+				panic(errors.New("invalid contract: the code must declare exactly one contract or contract interface"))
+			}
+
+			// The declared contract or contract interface must have the name
+			// passed to the constructor as the first argument
+
+			if declaredName != nameArgument {
+				panic(fmt.Errorf(
+					"invalid contract: the declaration must have the same name as the given name argument. epected %q, got %q",
+					nameValue.Str,
+					declaredName,
+				))
+			}
+
+			r.updateAccountContractCode(
+				runtimeInterface,
+				runtimeStorage,
+				declaredName,
+				code,
+				addressValue,
+				location,
+				checker,
+				contractType,
+				constructorArguments,
+				constructorArgumentTypes,
+				invocation.LocationRange.Range,
+				updateAccountContractCodeOptions{
+					createContract: !isUpdate,
+				},
+			)
+
+			codeHashValue := CodeToHashValue(code)
+
+			eventArguments := []exportableValue{
+				newExportableValue(addressValue, nil),
+				newExportableValue(codeHashValue, nil),
+				newExportableValue(nameValue, nil),
+			}
+
+			if isUpdate {
+				r.emitAccountEvent(
+					stdlib.AccountContractUpdatedEventType,
+					runtimeInterface,
+					eventArguments,
+				)
+			} else {
+				r.emitAccountEvent(
+					stdlib.AccountContractAddedEventType,
+					runtimeInterface,
+					eventArguments,
+				)
+			}
+
+			result := interpreter.DeployedContractValue{
+				Address: addressValue,
+				Name:    nameValue,
+				Code:    newCodeValue,
+			}
+
+			return trampoline.Done{Result: result}
+		},
+	)
+}
+
+type updateAccountContractCodeOptions struct {
+	createContract bool
+}
+
+// updateAccountContractCode updates an account contract's code.
+// This function is only used for the new account code/contract API.
+//
+func (r *interpreterRuntime) updateAccountContractCode(
+	runtimeInterface Interface,
+	runtimeStorage *interpreterRuntimeStorage,
+	name string,
+	code []byte,
+	addressValue interpreter.AddressValue,
+	location AddressContractLocation,
+	checker *sema.Checker,
+	contractType *sema.CompositeType,
+	constructorArguments []interpreter.Value,
+	constructorArgumentTypes []sema.Type,
+	invocationRange ast.Range,
+	options updateAccountContractCodeOptions,
+) {
+	// If the code declares a contract, instantiate it and store it
+
+	var contractValue interpreter.OptionalValue = interpreter.NilValue{}
+
+	createContract := contractType != nil && options.createContract
```

Comment:
Why do we check `contractType != nil` here -- is it ever valid for `contractType` to be `nil` in this function?

Reply:
Good question, I'll clarify this in a comment!

Reply:
@jordanschalm Tried to explain in 6c7148df4f4aac1ef705b08d272fe9ae34b5aa35, let me know if that makes sense, I can elaborate if not

Reply:
That helps a lot 👍